Add Grpc fault injection (#1108)
grafana/k6-docs: src/data/markdown/docs/40 xk6-disruptor/04 Examples/01 Inject Grpc faults into Service.md

---
title: 'Inject gRPC faults into Service'
excerpt: ''
---

> This example shows how [ServiceDisruptor](/javascript-api/xk6-disruptor/api/servicedisruptor) can be used for testing the effect of faults injected in the gRPC requests served by a service.

Reviewer: ```suggestion
This example shows a way to use the [ServiceDisruptor](/javascript-api/xk6-disruptor/api/servicedisruptor) to test the effect of faults injected in the gRPC requests served by a service.
```

> You will find the complete [source code](#source-code) at the end of this document. Next sections examine the code in detail.

Reviewer: ```suggestion
The complete [source code](#source-code) is at the end of this document. The next sections examine the code in detail.
```

The example uses [grpcpbin](https://grpcb.in), a simple request/response application that offers endpoints for testing different gRPC requests.

> The test requires `grpcpbin` to be deployed in a cluster in the namespace `grpcbin` and exposed with an external IP. the IP address is expected in the environment variable `GRPC_HOST`.

Reviewer: ```suggestion
The test requires `grpcpbin` to be deployed in a cluster in the namespace `grpcbin` and exposed with an external IP. The IP address is expected in the environment variable `GRPC_HOST`.
```

> You will find the Kubernetes manifests and the instructions of how to deploy it to a cluster in the [test setup](#test-setup) section at the end of this document. You can learn more about how to get the external IP address in the [expose your application](/javascript-api/xk6-disruptor/get-started/expose-your-application) section.

Reviewer: ```suggestion
For the Kubernetes manifests and the instructions on how to deploy it to a cluster, refer to the [test setup](#test-setup) section at the end of this document. To learn more about how to get the external IP address, refer to [Expose your application](/javascript-api/xk6-disruptor/get-started/expose-your-application).
```

Reviewer: Just trimming the language a bit, and capitalizing the title to indicate that it's a separate page.

## Initialization

> The initialization code imports the external dependencies required by the test. The [ServiceDisruptor](https://k6.io/docs/javascript-api/xk6-disruptor/api/servicedisruptor/) class imported from the `xk6-disruptor` extension provides functions for injecting faults in pods. The [k6/net/grpc](https://k6.io/docs/javascript-api/k6-net-grpc/) module provides functions for executing gRPC requests. We will also use the [check](https://k6.io/docs/javascript-api/k6/check/) function to verify the results from the requests.

Reviewer: ```suggestion
The initialization code imports the external dependencies required by the test. The [ServiceDisruptor](https://k6.io/docs/javascript-api/xk6-disruptor/api/servicedisruptor/) class imported from the `xk6-disruptor` extension provides functions for injecting faults in pods. The [k6/net/grpc](https://k6.io/docs/javascript-api/k6-net-grpc/) module provides functions for executing gRPC requests. The [check](https://k6.io/docs/javascript-api/k6/check/) function verifies the results from the requests.
```

```javascript
   if (__ENV.SKIP_FAULTS == "1") {
       return
   }
```

This code makes the function return without injecting faults if the `SKIP_FAULTS` environment variable is passed to the execution of the test with a value of "1". We will use this option to obtain a [baseline execution](#baseline-execution) without faults.

## Scenarios 

This test defines two [scenarios](https://k6.io/docs/using-k6/scenarios) to be executed. The `load` scenario applies the test load to the `grpcpbin` application for `30s` invoking the `default` function. The `disrupt` scenario invokes the `disrupt` function to inject a fault in the gRPC requests to the target application.

```javascript
    scenarios: {
        load: {
            executor: 'constant-arrival-rate',
            rate: 100,
            preAllocatedVUs: 10,
            maxVUs: 100,
            exec: "default",
            startTime: '0s',
            duration: "30s",
        },
        disrupt: {
            executor: 'shared-iterations',
            iterations: 1,
            vus: 1,
            exec: "disrupt",
            startTime: "0s",
        }
    }
 ```

 <Blockquote mod="note">

>  Notice that the `disrupt` scenario uses a `shared-iterations` executor with one iteration and one `VU`. This setting ensures the `disrupt` function is executed only once. Executing this function multiples times concurrently may have unpredictable results.

Reviewer: ```suggestion
 The `disrupt` scenario uses a `shared-iterations` executor with one iteration and one `VU`. This setting ensures the `disrupt` function is executed only once. Executing this function multiples times concurrently may have unpredictable results.
```

</Blockquote>

## Executions

<Blockquote mod="note">

> The commands in this section assume the `xk6-disruptor` binary is available in your current directory. This location can change depending on the installation process and the platform. Refer to the [installation section](/javascript-api/xk6-disruptor/get-started/installation) for details on how to install it in your environment.

Reviewer: ```suggestion
The commands in this section assume the `xk6-disruptor` binary is available in your current directory. This location can change depending on the installation process and the platform. Refer to [Installation](/javascript-api/xk6-disruptor/get-started/installation) for details on how to install it in your environment.
```

</Blockquote>

### Baseline execution

We will first execute the test without introducing faults to have an baseline using the following command:

<CodeGroup labels={["Linux/MacOS", "Windows PowerShell"]}>

```bash
xk6-disruptor --env SKIP_FAULTS=1 --env GRPC_HOST=$GRPC_HOST run grpc-faults.js
```

```PowerShell
xk6-disruptor --env SKIP_FAULTS=1 --env "GRPC_HOST=$Env:GRPC_HOST" run grpc-faults.js
```

</CodeGroup>

Notice the argument `--env SKIP_FAULT=1`, which makes the `disrupt` function return without injecting any fault as explained in the [fault injection](#fault-injection) section. Also notice the `--env GRPC_HOST` argument, which passes the external IP used to access the `grpcbin` application.

> You should get an output similar to the one shown below (click `Expand` button to see all output).

Reviewer: ```suggestion
You should get an output similar to the following (use the `Expand` button to see all output).
```

}

export const options = {
    scenarios: {
        load: {
            executor: 'constant-arrival-rate',
            rate: 100,
            preAllocatedVUs: 10,
            maxVUs: 100,
            exec: "default",
            startTime: '0s',
            duration: "30s",
        },
        disrupt: {
            executor: 'shared-iterations',
            iterations: 1,
            vus: 1,
            exec: "disrupt",
            startTime: "0s",
        }
    }
}
```
</Collapsible>

<Collapsible title="pb/hello.proto">

```
syntax = "proto2";

package hello;

service HelloService {
  rpc SayHello(HelloRequest) returns (HelloResponse);
  rpc LotsOfReplies(HelloRequest) returns (stream HelloResponse);
  rpc LotsOfGreetings(stream HelloRequest) returns (HelloResponse);
  rpc BidiHello(stream HelloRequest) returns (stream HelloResponse);
}

message HelloRequest {
  optional string greeting = 1;
}

message HelloResponse {
  required string reply = 1;
}
```

</Collapsible>

## Test setup

The tests requires the deployment of the `grpcbin` application. The application must also be accessible using an external IP available in the `GRPC_HOST` environment variable.

> The [manifests](#manifests) below define the resources required for deploying the application and exposing it as a LoadBalancer service.

Reviewer: ```suggestion
The following [manifests](#manifests) define the resources required for deploying the application and exposing it as a LoadBalancer service.
```

The style guidelines say to avoid directional llanguage, like "above" or "below", for accessibility: https://developers.google.com/style/accessibility

You can deploy the application using the following commands:

```bash
# Create Namespace
kubectl apply -f namespace.yaml
namespace/grpcbin created

# Deploy Pod
kubectl apply -f pod.yaml
pod/grpcbin created

# Expose Pod as a Service
kubectl apply -f service.yaml
service/grpcbin created
```

You must set the environment variable `GRPC_HOST` with the external IP address and port used to access the `grpcbin` service from the test script. 

> You can learn more about how to get the external IP address in the [expose your application](/javascript-api/xk6-disruptor/get-started/expose-your-application) section.

Reviewer: ```suggestion
Learn more about how to get the external IP address in the [Expose your application](/javascript-api/xk6-disruptor/get-started/expose-your-application).
```
